### PR TITLE
fix(stream): wire calibrated_logits_bytes_per_element into the VRAM pre-flight

### LINKED ISSUE #348: Layer streaming: calibrated_logits_bytes_per_element is dead code — wire it into estimate_stream_peak_vram
`src/soup_cli/utils/layer_stream.py` ships `calibrated_logits_bytes_per_element()`
(and the `measure_logits_loss_bytes_per_element()` probe underneath it), documented
as the guard against a future stack growing a fourth fp32 buffer in the loss path —
which would be under-budgeted by 12.5% today with nothing to catch it.

**It has no caller.** `grep -rn "calibrated_logits_bytes_per_element" src/` matches
only its own definition and the docstrings that reference it.
`trainer/stream_setup.py:455` calls `estimate_stream_peak_vram(...)` without
`logits_bytes_per_element=`, so the parameter is always `None` and the shipped
constant is always used. There is no CLI flag or env var that reaches it either.
The function is exercised only by `tests/test_issue327_logits_estimate.py`.

So the calibration is **an API, not a live path**. The hole it was written to close
is still open.

## Proposed fix

Thread it through: `_stream_budget_lines` calls `calibrated_logits_bytes_per_element()`
and forwards the result as `logits_bytes_per_element=` to `estimate_stream_peak_vram`.

The existing semantics already make this safe — the value is floored at the shipped
constant inside `estimate_logits_bytes`, so it can only *raise* the budget, never
lower it. A stack that measures below 12 is a probe that could not see the retention,
not a stack that does not pay it.

Two things to decide when implementing:

- **Cost.** The probe allocates `14 * vocab * max(tokens)` transiently — 96 MiB at
  the defaults — and needs CUDA. It returns `None` (falling back to the constant)
  rather than raising when it cannot run, which is the right shape, but it should
  probably be skipped when the fit decision cannot depend on it, the way
  `choose_tier` takes a callable so the ~9 s disk probe is only paid when it matters.
- **Reporting.** If the calibration raises the budget above the constant, the
  pre-flight panel should say so, with both numbers. A budget that silently differs
  from the documented constant is worse than no calibration.

## Acceptance criteria

- [ ] `_stream_budget_lines` forwards a calibrated value.
- [ ] Test asserting the pre-flight *uses* it: a stub probe returning a high value
      raises the prediction, and one returning a low value does not lower it.
- [ ] Test that a `None` probe (no CUDA) leaves the prediction byte-identical to
      today's.
- [ ] Panel line shows the calibrated value when it differs from the constant.

### PR DESCRIPTION
**Pull request by AmirF194**
## Root cause

`calibrated_logits_bytes_per_element()` exists to raise the VRAM pre-flight's budget when this stack's loss path measures a heavier retention than the shipped constant assumes: the guard against a future stack silently under-budgeting by 12.5% with nothing to catch it. It had no caller. `_stream_budget_lines` calls `estimate_stream_peak_vram()` without `logits_bytes_per_element=`, so the parameter was always `None` and the calibration ran only inside `tests/test_issue327_logits_estimate.py`.

## Fix

`_stream_budget_lines` now calls `calibrated_logits_bytes_per_element()` once and forwards the result to both `estimate_stream_peak_vram()` (the actual budget) and `estimate_logits_bytes()` (the panel's `logits` figure), so the two stay consistent. The value is floored at the shipped constant inside `estimate_logits_bytes`, so this can only raise the prediction, never lower it. When the calibration measures above the constant, the panel prints an extra line naming both numbers.

## Verification

- `tests/test_issue348_calibrated_logits_wiring.py` (new): a stubbed high calibration raises the predicted peak, a stubbed calibration below the constant does not lower it, a `None` probe (no CUDA) leaves the prediction byte-identical to today, and the panel adds a line only when the calibration diverges from the constant. Confirmed each assertion fails on `main` and passes on this branch.
- `tests/test_issue327_logits_estimate.py` (the existing calibration-arithmetic suite) passes unchanged.
- Full suite (`pytest tests/`, Docker, `pip install -e .[dev]`) on Python 3.10 and 3.12: 186 passed, 18 skipped (CUDA-only rows this sandbox has no GPU for). `ruff check` and `mypy` clean on the changed file.
- Not verified: the actual GPU probe path (`measure_logits_loss_bytes_per_element` with a real CUDA device) and the Windows/macOS legs of the CI matrix, since this sandbox is Linux CPU-only. The wiring is covered by stubbing the calibration function; the probe's own arithmetic was already pinned by #327's tests.

Fixes #348

### CODE REVIEW

**Review by MakazhanAlpamys — CHANGES_REQUESTED** [review]
Wiring is right, the floor is in the correct place, and forwarding one calibrated value to both the budget and the panel so they cannot disagree is the detail I would have asked for. CI is 11/11. Two things before merge, and the first is the same shape I have flagged twice now — which is worth saying plainly, because it is not carelessness, it is a specific blind spot worth naming once so it stops recurring.

## Blocking: the suite passes with the calibration disconnected from the budget

I removed the forwarding to `estimate_stream_peak_vram` and left it going to `estimate_logits_bytes` — i.e. the panel shows a calibrated logits figure while the budget that actually refuses a run keeps the shipped constant. That is precisely the defect #348 exists to close:

```
$ # estimate_stream_peak_vram(...)  <- logits_bytes_per_element removed
5 passed in 0.33s
```

Nothing failed. The reason is in the assertion:

```python
baseline = _budget_lines()[0]
...
raised = _budget_lines()[0]
assert baseline != raised
```

`_budget_lines()[0]` is one string carrying **both** numbers:

```
  peak VRAM    ~X.XX GB at batch 1 x seq 4 (logits Y.YY GB)
```

so a change in the `logits` half alone satisfies `baseline != raised`. The test pins *"the line changed"*, and the PR description claims *"a stubbed high calibration raises the predicted peak"* — those are different statements, and only the second is #348. The display figure is cosmetic; the budget is the thing that refuses a run.

The other two cannot cover it either: `test_a_below_constant_calibration_does_not_lower_the_peak` compares two equal outcomes, and `test_a_none_probe_leaves_the_peak_byte_identical_to_today` runs at `calibrated == 14 == LOGITS_BYTES_PER_ELEMENT`, where forwarding and not forwarding produce identical bytes by construction.

Assert on the peak alone, so the two halves are separable:

```python
def _peak_gb(line: str) -> str:
    return line.split("~", 1)[1].split(" GB", 1)[0]

assert _peak_gb(baseline) != _peak_gb(raised)
```

My mutation then fails, and — worth adding as a second assertion — the panel's `logits` figure moving *with* it becomes a separate claim rather than the one carrying the test.

**The pattern, said once.** #384: a positive test satisfied by `embed_tokens`/`norm`/`lm_head`, which never carried `.inner.`. #386: `override=0` pinned at config load but not at the resolver. Here: a peak assertion satisfied by the logits display. Each time the code is correct and the assertion is answered by something *adjacent* to the claim. The mutation test is what separates them — you ran it yourself on #386 and found the real value coming back, which is exactly the habit. Worth applying to the assertion as well as to the code: after writing it, ask what else could make it green.

## Second, and it needs your opinion rather than a fix: this makes an opt-in thing unconditional

`calibrated_logits_bytes_per_element()` now runs on **every** streamed run. #327's recorded decision was that the constant stays at 14 and only an **opt-in**, upward-only calibration was added. #348 asks for the wiring, so this is arguably the completion of that work rather than a reversal — but it is a change to a decision that was made deliberately, and it should be made out loud rather than as a side effect.

The concrete cost, from the probe's own docstring: one transient allocation of `14 * vocab_size * max(tokens)` — **96 MiB at the defaults** — plus two `torch.cuda.synchronize()` calls, on the card that is short of VRAM, before the fit decision has been taken. It also calls `torch.cuda.reset_peak_memory_stats()`, so anything downstream reading `max_memory_allocated` now measures from the probe rather than from process start.

Today it is a no-op in effect: `measured` is 12.0 with zero spread on this stack, `+2` retention gives exactly 14, `max(14, 14)` is 14, and the extra panel line correctly does not print. So the cost buys nothing until the stack changes — which is the entire point of a guard, and I am not arguing against it.

I would land it as written and take the cost. But say which you intend, so the next reader of #327's decision finds the reason: **(a)** always-on, cost accepted, `#327`'s "opt-in" wording updated in the same PR; or **(b)** gated behind a config field, default off, with the guard reachable for anyone who wants it. If (a), one line in the CHANGELOG entry naming the 96 MiB is enough.

---

The test change is the only thing I need. Answer (a)/(b) in a sentence and I will merge on your next push.

**Comment by AmirF194** [comment]
Fixed. Split the peak GB out of the line and asserted on it alone (`_peak_gb`), then added a second test asserting the logits figure separately so each half is its own claim. Ran the mutation you described (drop `logits_bytes_per_element=` from the `estimate_stream_peak_vram` call only) in a clean container: the new peak assertion fails on it, the old one did not.

Answer to the second point: (a), land as written, cost accepted. Updated #327's entry to drop the "opt-in" wording and named the 96 MiB / two-sync cost in #348's entry.

Pushed to the same branch.

**Comment by MakazhanAlpamys** [comment]
Both closed, and the second one better than I asked.

**The peak assertion is now real.** Ran the mutation again — calibration reaching the panel but not the budget:

```
FAILED ...::TestCalibrationIsForwardedToTheBudget::test_a_high_calibration_raises_the_predicted_peak
1 failed, 5 passed
```

Exactly the right test, and the five others unaffected. Splitting `_peak_gb` out and giving the logits figure its own test is the shape that stops the two halves covering for each other, and you checked the old assertion did *not* fail on the mutation — which is the half people skip.

**You answered (a) and then did the part that makes it a decision rather than a side effect:** dropping "opt-in" from #327's entry and naming the 96 MiB + two-sync cost in #348's, with the honest note that it is a no-op in effect today. Someone reading #327 in six months now finds a reason instead of a contradiction. I asked for one sentence; the CHANGELOG edit is worth more than the sentence.

CI 11/11. Merging.

---

Three PRs in four days, and the through-line is worth naming since it is the reason these went fast: each time the code was right first try, and what needed another round was an *assertion* that something adjacent could satisfy. You closed that gap yourself this time, before I could — you ran the mutation, checked the old test did not catch it, and only then wrote the new one. That is the whole technique.

#348 closes with this. If you want more, `good first issue` and `help wanted` are [labelled in the tracker](https://github.com/MakazhanAlpamys/Soup/labels/help%20wanted); nothing is reserved and there is no expectation either way.

**Comment by AmirF194** [comment]
Thanks. Running the mutation before writing the new assertion is the part I want to keep doing on every test I write from now on, not just when a review points it out.
